Adds a new Contacts page to the Payments section (#669)
BitcoinDesign/Guide: guide/payments/contacts.md

---
layout: guide
title: Contacts
description: Design reference for a contacts feature in bitcoin wallets.
nav_order: 4
parent: Payments
permalink: /guide/payments/contacts/
main_classes: -no-top-padding
image: /assets/images/guide/payments/activity/contacts-preview.jpg
image_base: /assets/images/guide/payments/contacts/
imagesAddContact:
    - file: manual-add-empty
      alt: Empty contact list screen

>       caption: Initially, a walllets contact list is empty. Importing them from the device can be made available.

Reviewer: ```suggestion
      caption: Initially, a walllet's contact list is empty. Importing them from the device can be made available.
```

    - file: manual-add-contact-add-address
      alt: Screen asking the user to enter a bitcoin address
      caption: Users can manually enter addresses to add.
    - file: manual-add-contact-add-address-details
      alt: Address entry screen with expanded information about supported addresses
      caption: Information about supported address formats should be readily available.
    - file: manual-add-contact-add-address-valid
      alt: Address entry screen with inline validation for a Lightning address
      caption: Inline validation lets users know about if addresses are accepted.
    - file: manual-add-contact-add-address-valid-on-chain
      alt: Address entry screen with inline validation for an on-chain address
      caption: Particular properties of address types can be pointed out right away.
    - file: manual-add-contact-with-address
      alt: A contact with a Lightning address assigned
      caption: A contact with a Lightning address associated.
    - file: manual-add-contact-with-on-chain-address
      alt: A contact with an on-chain address assigned

>       caption: A contact with an on-chain. Once a paymen is made to the address, it disappears from this list.

Reviewer: ```suggestion
      caption: A contact with an on-chain address attached. Once a payment is made to the address, it disappears from this list.
```

imagesHome:
    - file: sending-home
      alt: Home screen with amount input
      caption: The user enters an amount and taps "Pay".
    - file: sending-pick-contact
      alt: Screen with a list of contacts to choose from
      caption: Next, the user is asked to choose a contact to send to.
    - file: sending-review
      alt: Invoice screen with a contact assigned
      caption: Final review of the payment before sending it.
imagesReceive:
    - file: receive-home
      alt: Home screen with amount input
      caption: The user enters an amount and taps "Request".
    - file: receive-qr
      alt: Payment request screen
      caption: An invoice screen with minimal information.
    - file: receive-qr-with-contact
      alt: Payment request screen with a contact assigned
      caption: The same invoice with a contact assigned.
imagesActivity:
    - file: activity
      alt: Activity screen showing transactions with minimal information
      caption: By default, payments contain little to no contextual information.
    - file: activity-annotated
      alt: Activity screen showing transactions with assigned contacts

>       caption: Assigning contacts to payment makes them much easier to understand and work with.

Reviewer: ```suggestion
      caption: Assigning contacts to payments makes them much easier to understand and work with.
```

    - file: activity-transaction
      alt: Transaction details
      caption: Transaction detail screens should have convenient options to assign a contact.
    - file: sending-pick-contact
      alt: Screen with a list of contacts to choose from
      caption: The user picks a contact in a slide-up modal.
    - file: activity-transaction-annotated
      alt: Transaction details with assigned contact
      caption: A payment with an assigned contact.

---

{% include picture.html
    image = "/assets/images/guide/payments/contacts/contacts-header.jpg"
    retina = "/assets/images/guide/payments/contacts/contacts-header@2x.jpg"
    mobile = "/assets/images/guide/payments/contacts/contacts-header-mobile.jpg"
    mobileRetina = "/assets/images/guide/payments/contacts/contacts-header-mobile@2x.jpg"
    alt-text = "Illustration of large contact cards"
    width = 1600
    height = 400
    layout = "full-width"
%}

# Contacts

<!--

-->

Whether we’re sending emails, physical mail, or following someone on social media, we primarily think in terms of names and faces, and not the respective address or user ID.

> Invoices, node IDs and other transaction endpoints in bitcoin and Lightning are highly unintuitive. Abstracting them via a contact list can create a much smoother user experience. There are many [payment request formats]({{ '/guide/payments/request/payment-request-formats/' | relative_url }}), each with unique properties and varying levels of maturity and adoption, requiring unique design solutions. This page uses the more approachable term "address", along with various UI techniques, to abstract these complexities for users.

Author: I don't see a problem with the #671 page, as it talks about invoices. My UI proposal does not mix invoices with addresses.

It takes some squinting to accept the term "address", for sure. Just like with "street address", "web address", and "email address". There's also "web address" and "IP address", which are pretty different things, but you can type both in the same input field in the browser and get where you want to. That makes "address" a very flexible, and overloaded, term. I just think we have to forget about technical lingo here. Can I repeatedly send something to this weird-looking identifier? Then it's an address (no matter if it's a Lightning address, zPub, offer, etc).

Destination is not quite right, as I can also receive from certain payment request formats. If I save your Lightning address, and you send from that node, it should also be linked to that payment. I know it's not 100% the same thing, as the Lightning address is not baked into the invoice, but it's the same from the user perspective.

This is such a squishy topic that can easily result in totally different outcomes. When I looked at it in the [work stream](https://youtu.be/GPewV_C-w2Q) last week, I basically decided to declare bankruptcy on the technical terminology and trying to find the perfect term, and just go with something that is widely accepted online. Just had to squint. 🤷🏻‍♀️

> Invoices, node IDs and other transaction endpoints in bitcoin and Lightning are highly unintuitive. Abstracting them via a contact list can create a much smoother user experience. There are many [payment request formats]({{ '/guide/payments/request/payment-request-formats/' | relative_url }}), each with unique properties and varying levels of maturity and adoption, requiring unique design solutions. This page uses the more approachable term "address", along with various UI techniques, to abstract these complexities for users.

Reviewer 2: Hmm. Ok, that argument is somewhat convincing to me. 

It looks very clean in your UI designs which simply call it "address", and then the user can expand to see the different formats. I'm curious what happens if using the word in this manner begins to seep out into regular usage. Do people start appending the term "bitcoin address" onto it? They likely would, in order to differentiate from "web address", "email address", etc. Not technically accurate, but I can live with it. We can easily use "on-chain address" to clarify in situations where a bitcoin address is needed.

It would be good to get some other opinions on it, but I'm warming up to the idea.

> Invoices, node IDs and other transaction endpoints in bitcoin and Lightning are highly unintuitive. Abstracting them via a contact list can create a much smoother user experience. There are many [payment request formats]({{ '/guide/payments/request/payment-request-formats/' | relative_url }}), each with unique properties and varying levels of maturity and adoption, requiring unique design solutions. This page uses the more approachable term "address", along with various UI techniques, to abstract these complexities for users.

Reviewer 3: Mmh yeah I think it's starting to warm up on me, I don't see any glaring issues with it! Makes a lot of sense throughout the section

Let's go over common user interactions around contacts. This will illustrate how such a feature could work, and helps explain the underlying design problems and decisions.

### Adding a contact

The most basic interaction is that a user manually adds a contact by entering their name and an address. This is likely not the most common user flow, as most contacts will be added to incoming payment requests, as illustrated further below. But wallets should generally support manual options, as users may not be able to avoid them.

> {% include image-gallery.html pages = page.imagesAddContact %}

Reviewer 2: I like the idea of the being able to cross-link the contact metadata with payment details (addresses, as they are referred to on this page). However, there's a lot of stuff I find really strange about this series of example screens.

- xPubs - why can the user import an xPub to a contact? I'm very skeptical that this feature is a good design decision for a mobile LN wallet.
  - **tl;dr** I don't think the majority of mobile users will benefit from storing xPubs with their contacts, and it could actually be a net harm if they mess something up.
  - We all know address re-use has privacy implications. So in theory, you could give your friend your xPub and they can generate fresh addresses to coins to you.
  - But is this a good idea? What if two friends have my xPub? What if two people send to the same address? Are my friends scanning the blockchain to see if any of my addresses have been used already? Are their wallets constantly scanning the blockchain for every single xPub of every contact in their list? (Reductio ad absurdum)
  - Like I brought up on the Jam call this week, I really don't think users should be in the habit of sending the most valuable monetary asset in the entire world to addresses, invoices, etc. that the receiving party has not consented to receiving from.
    - Yes, it's technically possible to send to _any address you want to_, but not good practice, IMO.
  - To me, the most useful situation for things like xPubs for most users are things like BTCPayServer where you want to the instance to generate receive addresses without having your private keys, or maybe something like auto-withdrawals from Swan Bitcoin.
- On-chain addresses
  - Why is there so much focus on on-chain addresses in these screens?
  - You already mentioned in the captions that the address can only be used once. Ok, perfect, I agree, that's a nice feature to restrict the user from re-using an address.
  - But tell me -- if these addresses can only be used once, what user is going to be manually copy and pasting on-chain addresses into their contacts list so they can use those addresses only once?
  - It looks incredibly slick, but copy/pasting individual addresses seems like a really obtuse user experience.
  - I'm not saying no user would ever want that feature -- I just don't see 99% of users doing that, and it seems weird to be featuring that so prominently at the very top of this page.

The ability to cross-link a contacts list with one's on-chain transaction history is very useful. However, I don't really think the ability to manually add on-chain data to a contacts list is a feature that's going to help most users. There's a subtle but important distinction between those two use cases.

Suggested remedy: cut out the manually adding an on-chain address screens or swap it out with a BOLT12 offer or LNURL pay request.

Author: Totally understand your rationale and I think some tweaks are necessary. You are viewing these flows very different than what I intended them to be.

Here are some thoughts I had in mind with this:

1. Wallets should be able to identify all address types in existence and have an appropriate response for the user with practical information (even if it has to state that the format is not supported/recommended).
1. We cannot predict the scenarios users find themselves in. There might be really good reasons to save an on-chain address or xpub. I'd rather have the app inform users appropriately (via point 1) and let them go through with it (maybe with a bit of extra friction) than completely restrict it. If we artificially restrict something like this, users may be stuck and won't be able to access their funds at an important moment.
1. Payment request formats are a moving target, so I was more focused on explaining a good UI system to handle various scenarios rather than describing which ones to use for the daily spending use case. Showing two different formats with very different properties helps with that.
1. This first user flow is meant to guide the reader through the most basic elements covered on the page. It is intentionally completely manual and a bit drawn out, because everything else builds on it. Especially with on-chain addresses, the clipboard flow is much more likely to occur. But the idea was to start with the manual basics to avoid instant overload (the first iteration of this page took that approach).

What I'd like to do as a response:

1. Add an example of manually adding an xpub, with a warning below ("You can make repeated payments to this address, but it is highly recommended not to use it. This address type (extended public key) can cause privacy and security problems if not carefully managed"). Maybe add a check-box to let users confirm that they understand the implications. In the end, designers need to consider what their app supports and carefully think through how their app informs users. Showing outlier examples, along with "good" examples helps with that.
2. Add a note below that the app should emphasize Lightning formats for this particular use case (summarizing your considerations above).

How does that sound?

Author: OK, did something a little different. I split out the various states for the "add address" interaction into a separate paragraph. This allows the first paragraph to be just about the user flow of contact and address input. The second paragraph goes into the discussion of supporting formats and UX considerations.

I added two states for xpubs. One is a UI that allows their use, but the user needs to toggle an "I understand" box. The second one disallows their use. [Check it here](https://deploy-preview-669--sad-borg-390916.netlify.app/guide/payments/contacts/#supporting-various-payment-request-formats). Also tried to clarify that designers need to consider this for their own apps. It's that middle ground again in the guide between showing the exact solution for a single use case and laying out alternatives for designers to lay out the thought process and explain some of the underlaying problems.

Reviewer 2: @GBKS I love it, thanks for considering my very wordy feedback.

Yeah, I agree with with you that edge cases must be considered. I don't want to argue that apps should _never_ support some of the features you describe in this PR; my primary concern was how it looks presenting it so close to the top of the page. I feel that your new updates have clarified things a lot.

    - file: activity-annotated
      alt: Activity screen showing transactions with assigned contacts
      caption: Assigning contacts to payments makes them much easier to understand and work with.
    - file: activity-transaction
      alt: Transaction details
      caption: Transaction detail screens should have convenient options to assign a contact.
    - file: sending-pick-contact
      alt: Screen with a list of contacts to choose from
      caption: The user picks a contact in a slide-up modal.
    - file: activity-transaction-annotated
      alt: Transaction details with assigned contact
      caption: A payment with an assigned contact.

---

{% include picture.html
    image = "/assets/images/guide/payments/contacts/contacts-header.jpg"
    retina = "/assets/images/guide/payments/contacts/contacts-header@2x.jpg"
    mobile = "/assets/images/guide/payments/contacts/contacts-header-mobile.jpg"
    mobileRetina = "/assets/images/guide/payments/contacts/contacts-header-mobile@2x.jpg"
    alt-text = "Illustration of large contact cards"
    width = 1600
    height = 400
    layout = "full-width"
%}

# Contacts
{:.no_toc}

<!--

> 

Reviewer 3: Although I know the images are used from the UI kit it would still be handy to have the link to the Figma file

Author: Added.

-->

Whether we’re sending emails, physical mail, or following someone on social media, we primarily think in terms of names and faces, and not the respective address or user ID.

Invoices, node IDs and other transaction endpoints in bitcoin and Lightning are highly unintuitive. Abstracting them via a contact list can create a much smoother user experience. There are many [payment request formats]({{ '/guide/payments/request/payment-request-formats/' | relative_url }}), each with unique properties and varying levels of maturity and adoption, requiring unique design solutions. This page uses the more approachable term "address", along with various UI techniques, to abstract these complexities for users.

Let's go over common user interactions around contacts. This will illustrate how such a feature could work, and helps explain the underlying design problems and decisions.

---

<div class="glossary-toc" markdown="1">
* Table of contents
{:toc}
</div>

---

> 

Reviewer 3: We need to have a section on the wallet users contact card which I would put first here. This would cover how they can add their own addresses to it and so forth. Maybe it comes with an address already (say if they setup a Lightning address during on boarding). Likely want to cover flows around sharing their contact card as that is likely what many will do with theirs as well.

Author: Having a contact card for the owner doesn't seem quite right. When "sharing your contact", don't you want to use the request flow? If a wallet doesn't let you add a Lightning address, would the screen be empty?

Reviewer 3: > Having a contact card for the owner doesn't seem quite right. 

I don't know of any app that uses contacts that does not let the user have their own contact card. 

> When "sharing your contact", don't you want to use the request flow?

Not necessarily, if I have my own contact card and it has say 3 addresses in it I'd rather just share that contact card with someone and they can choose one of the 3. 

How would you share your Lightning address during the request flow? 

> If a wallet doesn't let you add a Lightning address, would the screen be empty?

It would  just start empty if the app didn't have the user setup some kind of address during onboarding or something. It could still just be blank and have the option to add a profile image.

### Adding a contact

> The most basic interaction is that a user manually adds a contact by entering their name and an address. This is likely not the most common user flow, as most contacts will be added to incoming payment requests, as illustrated further below. But wallets should generally support manual options, as users may not be able to avoid them.

Reviewer 3: ```suggestion
The most basic interaction is that a user manually adds a contact by entering their name and an address. This is likely not the most common user flow, as most contacts will be created with incoming payment requests, as illustrated further below. But wallets should generally support manual options, as users may not be able to avoid them.
```

> {% include image-gallery.html pages = page.imagesAddContact %}

Reviewer 3: Adding one more state around when a user enters an unrecognized or unsupported address type would be handy.

Say if the user enters a Lighting address or Taproot on-chain address but the wallet does not support sending to these yet this should be clear here.

### Supporting various payment request formats

> Based on your use case, your application may choose to only support certain payment request formats. It should still recognize all formats, and provide the user with appropriate feedback. You may also choose to add warnings before permitting use of certain formats if there are privacy and security risks the user should be aware of.

Reviewer 3: ```suggestion
Based on your use case, your application may choose to only support certain payment request formats. It should still recognize all formats, and provide the user with appropriate feedback. You may also choose to add warnings before permitting use of certain formats if there are privacy and security risks the user should be aware of.

Only [reusable payment requests]({{ '/guide/payments/request/#reusable-payment-requests' | relative_url }}) should be possible to be imported by users. 
```

Author: I assume by `import`, you mean `save to contact for future use`. That sentence goes counter to the first screen in the gallery, which shows a single-use on-chain address.

Reviewer 3: > I assume by import, you mean save to contact for future use.

Yep!

> That sentence goes counter to the first screen in the gallery, which shows a single-use on-chain address.

Personally I'd just remove the options around saving an on-chain address (and xpubs) to a contact and stick with reusable payment requests only. On-chain payments should be more intentional so the generating of a new address each time isn't a big burden

Author: I want to keep them, since this section is all about how to highlight various properties of different formats in the UI, so users can work with them without having to understand anything technical. The text is clear in that designers should think deeply about this and decide what to include based on their use cases.

> {% include image-gallery.html pages = page.imagesSupportedFormats %}

Reviewer 3: Adding one more state around when a user enters an unrecognized or unsupported address type would be handy.

Say if the user enters a Lighting address or Taproot on-chain address but the wallet does not support sending to these yet this should be clear here.

Author: Added.

### Contact editing

The contacts screen should offer various features for editing and organization. Also see [Requesting bitcoin &middot; Contact cards]({{ '/guide/payments/request/#contact-cards' | relative_url }}).

> {% include image-gallery.html pages = page.imagesMultiAddressContact %}

Author: How about those optional labels I added in the third screen in this gallery?

`Offer` and `Node ID` are probably also not too helpful for users.

> {% include image-gallery.html pages = page.imagesMultiAddressContact %}

Reviewer 3: Yeah they are better than just displaying the string, users will probably have to know what the different addresses are (as they offer unique features) so Offer and Node ID aren't that bad

> {% include image-gallery.html pages = page.imagesMultiAddressContact %}

Author: Goal is that they don't have to understand the address types, but that the UI highlights what users need to know at the time it is relevant.

### Importing an address

In this scenario, the user has copied a Lightning address to the clipboard and opened the app. The clipboard is automatically scanned, the app identifies the addres and offers the user appropriate options. Here, the user adds the address as a new contact.

> {% include image-gallery.html pages = page.imagesImportAddress %}

Reviewer 3: The add contact button should prompt if the user wants to add to an existing contact or create a new contact.

Author: Good point. [Daniel had some feedback](https://www.figma.com/file/VB3GQdAnhl8yta44DY3PSV/Bitcoin-UI-Kit?node-id=3202%3A83932) on changing this modal to just show three buttons. This would simplify the overall effort by re-using existing flows (which would address your point, as the contacts modal has a "+" button).

Reviewer 3: LGTM :)

This flow can also be initiated by scanning a QR code or tapping a payment link. For more, see [Sending bitcoin &middot; Payment entry points]({{ '/guide/payments/send/#payment-entry-points' | relative_url }}).

> ### Importing an invoice

Author: `Assigning a payment` is too close to the `Adding a contact to a completed payment` section. The current title is a better match for the page as a whole.

Here, the user has scanned a payment request and assigns a contact to the payment.

> Some payment request formats may include an address that can receive payments repeatedly. In this case, the address is added to the contact for future use. Otherwise, only the payment is linked.

Reviewer 3: ```suggestion
Some payment request formats may include an address that can [receive payments repeatedly]({{ '/guide/payments/request/#reusable-payment-requests' | relative_url }}). In this case, the address is added to the contact for future use. For [single use payment requests]({{ '
/guide/payments/request/#single-use-payment-requests' | relative_url }}), only the payment is linked.
```

> Invoices may also contain recipient names. These can be used to suggest the name for a new contact to the user. Names can be spoofed, so they should not be automatically assigned without user approval.

Reviewer 3: ```suggestion
Payment requests may also contain recipient names. These can be used to suggest the name for a new contact to the user. Names can be spoofed, so they should not be automatically assigned without user approval.
```

{% include image-gallery.html pages = page.imagesPayInvoice %}

### Sending from the home screen

Initiating a payment can be as simple as entering an amount and tapping a contact that has a re-usable address associated. Otherwise, an extra step is needed to manually enter a destination address.

{% include image-gallery.html pages = page.imagesHome %}

> ### Adding a contact to an outgoing invoice

Reviewer 3: ```suggestion
### Adding a contact to a shared payment request 
```

Contacts can also be assigned to invoices that will be paid by others.

{% include image-gallery.html pages = page.imagesReceive %}

> ### Adding a contact to a completed payment

Reviewer 3: Seems to be an overlap with this section and the Importing an invoice section.

Author: Very different flows. One is when importing an invoice, making a payment an assigning the contact right in that flow. The other is is going back to completed payments and adding the contact.

Reviewer 3: But you don't import an invoice, invoices are single use, you're still importing an address (as you've defined them).

When browsing the [activity]({{ '/guide/payments/activity/' | relative_url }}) screen, users may come across payments they want to assign contacts to. While applications should try to automatically make connections between payments and contacts by matching addresses, this is not possible oftentimes.

{% include image-gallery.html pages = page.imagesActivity %}

> 

Author: Could be just a small mention, don't think it needs a whole section.

> 

Author: Added at the bottom of the page.

    alt-text = ""
    width = 1600
    height = 400
    layout = "full-width"
%}

# Contacts

<!--

-->

Whether we’re sending emails, physical mail, or following someone on social media, we primarily think in terms of names and faces, and not the respective address or user ID. Invoice IDs and other identifiers in bitcoin and Lightning are highly unintuitive, and abstracting them via a contact list can create a much smoother user experience. However, there are many [payment request formats]({{ '/guide/payments/request/payment-request-formats/' | relative_url }}), each with unique properties, requiring unique design solutions.

### The basics

<div class="center" markdown="1">

{% include image.html
   image = "/assets/images/guide/payments/contacts/contact.png"
   retina = "/assets/images/guide/payments/contacts/contact@2x.png"
   alt-text = ""
   width = 250
   height = 541
   layout = "float-right-desktop -background -shadow"
%}

On the surface, a contact entry represents a person, business or other entity that the user sends payment to, or receives payments from.

- A name
- An image (profile, avatar, company logo…)

> - A list of identifiers

Reviewer 3: ```suggestion
- A list of reusable payment requests
```
I still think this term identifiers is too jargony

- Options
- Prior interactions
- Organizational features like tags & favorites

Typical user flows involving contacts are:

- Choosing a contact to send bitcoin to

> - Assigning a contact to a generated payment request

Reviewer 3: ```suggestion
- Assigning a contact to a generated [reusable payment request]({{ '/guide/payments/request/#reusable-payment-requests' | relative_url }})
```

- Browsing the wallet payment history
- Organizing contacts (merging duplicates, deletion, etc)

</div>

Let’s go over each of these.

> 

Reviewer 3: Could add a section on the Contact list and how to organize it's UI.

A second aspect of this is that these names can be easily spoofed, there is no guarantee that they are accurate. So wallets should parse these identifiers and suggest them to users for their approval.
Identifiers
When we think of traditional contacts, we think of phone numbers, emails, and physical addresses. These identifiers are typically static and reliable for long periods of time. This is different in bitcoin, as each payment has its own address or invoice with a unique identifier. There are several approaches to re-usable identifiers (payment request formats), each with its own advantages and disadvantages.

##### On-chain identifiers

<div class="center" markdown="1">

{% include image.html
   image = "/assets/images/guide/payments/contacts/on-chain-identifiers.png"
   retina = "/assets/images/guide/payments/contacts/on-chain-identifiers@2x.png"
   alt-text = ""
   width = 250
   height = 116
   layout = "float-right-desktop -background -shadow"
%}

> There are no good options for re-usable identifiers for on-chain payments, primarily due to privacy concerns.

Reviewer 3: PayNym?

Reviewer 3: Works fine with Samourai, what if a product designer interested in creating a privacy focused application reads the guide - they would want to be aware of something like this. We could have a disclaimer saying it's not an official BIP but it could still be used if your user base if some product stacks if it caters to their target audience. We talk about BOLT12 and other non-standards throughout the guide, what's the difference with this? 

> Unanimously Discourage for implementation

Where is this from? Say's its still in draft mode on the bips GitHub repo - https://github.com/bitcoin/bips. Unanimously by who exactly?

Reviewer 4: I don't know but it seems we should find out what the concerns were before promoting its usage in the guide. Samourai's response to you on Twitter wasn't exactly enlightening. We aren't "seeking permission" we're "seeking understanding in order to be responsible."

Author: [Reading up](https://github.com/Multibit-Legacy/multibit-hd/issues/818#issuecomment-161673066) on this a bit, here's my understanding:
- Using a payment code requires the sender to create a notification transaction to the receiver, which establishes a log of this relationship (forever)
- Both parties generate a new wallet for this relationship
- The sender can generate addresses to send to
- The receiver generates addresses to observe for incoming transactions

Problems brought up:
- Notification addresses are spam. Since they have to be created between every two unique sender-receiver pair, this could get pretty extensive
- Wallet software needs to manage lots more wallets (dozens?) and observe even more addresses (hundreds, thousands?)

Counterpoint from [SamouraiDev](https://twitter.com/SamouraiDev/status/1345346094692249601?s=20) is that notification transactions are not needed if a good meta data scheme is used. Not sure how that works though. If a wallet is not notified of someone generating addresses and sending transactions, how does it know that they even exist? Should it just scan all transactions forever and match it against a million generated wallets and addresses to see if something is there?

As we've discussed a few times, including the jam session yesterday, how valuable is having a short identifier as a feature? What are the use cases? Are they important enough to warrant this overhead?

Author: [One more](https://twitter.com/LukeDashjr/status/1495978585857073154?s=20), somewhat short and cryptic.

Reviewer 5: @GBKS The point of the notification transaction is that the "payment channel" connection is very easily recoverable from the blockchain. It's a tradeoff but it's not necessary. And definitely not a spam. If you don't make notification transaction, you can come up with any other scheme on how the recipient should be notified so they can create a shared secret, derive addresses and listen for incoming transactions.

"unanimously discouraged" stamp of disapproval is irrelevant because BIP47 has been working on-chain on mainnet for years and more wallets are implementing it now.

Author: @pajasevi thanks for the details. Which other wallets are implementing this?

In terms of creating a "closed system", the notification transaction seems to be a pretty essential part as it works both for automation/communication and for archival.

Now I wish that "unanimously discouraged" always came with a rationale.

Reviewer 5: Apart from Samourai, Sparrow wallet already implemented it, some guys are working on an implementation for BlueWallet (for bounty) and I believe there might be a third as well but can't recall which one. Also, there are BIP47 open-source libraries written in Java, Javascript and Rust for others to use.

RE: notification tx - wallet MUST always listen for incoming notification TXs but it CAN also have an ability to accept notification via some other method. In this case, it won't be recoverable of course and that should be reflected in the UI.

Author: Took a quick look at this. Here's the [BlueWallet bounty](https://github.com/BlueWallet/BlueWallet/issues/2883).

My impression is that people are mixing up BIP 47 and PayNyms. From the bounty issue:
> Having more wallets support this venmo style method for sending money to your friends, would benefit the ecosystem in general.

Aren't PayNyms a centralized solution by Samourai that gets you nice Twitter-like handles ("+christoph"). [Here's a line](https://github.com/sparrowwallet/sparrow/blob/ce6b371206b1609462199ee26488e657e6b4251d/src/main/java/com/sparrowwallet/sparrow/paynym/PayNymService.java#L55) in the Sparrow codebase that uses the PayNym API (`https://paynym.is/api/`). If so, then I don't think it's a good idea to push this as it's completely centralized.

Reviewer 3: > If so, then I don't think it's a good idea to push this as it's completely centralized.

Playing devils advocate here but - I'd put as having the same tradeoffs as Lightning addresses, trusted but still non-custodial. If we go with that logic why do we recommend Lightning addresses? 

If we introduce any content around this in the guide we should just refer to it as payment codes / BIP47 and not PayNym (Samourai's implementation of the aforementioned). We don't have to recommend to use it or not use it but can just have it mentioned. Product developers can decide if they want to use it not.

Reviewer 5: Paynyms are an implementation of BIP47 payment codes. paynym.is is just a directory of payment codes with generated short name and image for more user-friendly usage. A wallet implementing BIP47 properly doesn't have to use this directory and still will be completely interoperable with other implementations.

Author: @pajasevi that's my understanding as well. I just don't think that's clear to the people in the Blue Wallet bounty thread.

@Bosch-0 Lightning addresses rely on servers (for the underlying LNURL service) but are not centralized. Your Lighting address and my Lightning address can be hosted on different servers. Not so with +paynym handles.

Let's see how we can get BIP 47 info added. Will try to think through that.

- An on-chain bitcoin address can technically be reused and receive multiple payments. This is highly discouraged, since it allows participants and observers to trace the activity.
- Extended public keys allow for the generation of many on-chain addresses. A user could share this key with others so they can generate addresses as needed. This is also discouraged, as anyone with this key can observe all transactions in the wallet.

</div>

> ##### Lightning identifiers

Reviewer 3: As we are going with Lightning first, we should discuss these before on-chain. I'd just combine these two sections into one "Reusable payment requests" section that covers both.

<div class="center" markdown="1">

{% include image.html
   image = "/assets/images/guide/payments/contacts/lightning-identifiers.png"
   retina = "/assets/images/guide/payments/contacts/lightning-identifiers@2x.png"
   alt-text = ""
   width = 250
   height = 273
   layout = "float-right-desktop -background -shadow"
%}

> Lightning invoices expire and can only be used once. However, Lightning nodes are always online, and can generate and serve invoices on-demand.

Reviewer 3: Not true when discussing mobile lightning nodes (main focus of this section).

> - Static IP or onion addresses of Lightning nodes could serve as reliable identifiers. However, this requires wallets to support [Keysend](https://lightning.readthedocs.io/lightning-keysend.7.html).

Reviewer 3: ```suggestion
- Lightning node ID's can serve as reliable identifiers. However, this requires wallets to support [Keysend](https://lightning.readthedocs.io/lightning-keysend.7.html).
```

- [BOLT12 offers](https://bolt12.org) (a draft spec) include the information required to contact a Lightning node to serve an invoice. This communication happens over the Lightning network.
- [BOLT12 addresses](https://github.com/rustyrussell/bolt12address) (a draft spec) are an abstraction of BOLT12 offers using the familiar format of email addresses.
- LNURL-Pay is similar to BOLT12 offers, but communication happens over the web and requires a server.
- Lightning Addresses are similar to BOLT12 addresses, but are layered on top of LNURL-Pay.

As we’ve now seen, there are no great options on-chain, and most of the techniques around Lightning are still either in the specification phase, or not widely implemented in standardized, convenient ways. While this will change over time, wallets are currently limited in how convenient contact features can be made.

</div>

### User flows

##### Choosing a contact to send bitcoin to

> In an ideal flow, a user simply taps a contact and enters an amount to make a payment. Due to the complexities around identifiers outlined above, a more likely scenario is that users need to manually assign a contact to an invoice or completed payment. This should be made as convenient as possible.

Reviewer 3: ```suggestion
In an ideal flow, a user simply taps a contact and enters an amount to make a payment. Due to the complexities around identifiers outlined above, a more likely scenario is that users need to manually assign a contact to an reusable payment request or completed payment. This should be made as convenient as possible.
```
